Kernel/Devices/AHCIPort.cpp

```diff
@@ -0,0 +1,438 @@
+*    and/or other materials provided with the distribution.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+* DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <Kernel/Devices/AHCIChannel.h>
+#include <Kernel/Devices/AHCIPort.h>
+#include <Kernel/VM/MemoryManager.h>
+
+#define FIS_TYPE_REG_H2D 0x27
+#define FIS_TYPE_PIO_SETUP 0x5F
+
+#define AHCI_DEV_CLB 0x00
+#define AHCI_DEV_CLBU 0x04
+#define AHCI_DEV_FB 0x08
+#define AHCI_DEV_FBU 0x0C
+#define AHCI_DEV_IS 0x10
+#define AHCI_DEV_IE 0x14
+#define AHCI_DEV_CMD 0x18
+#define AHCI_DEV_TFD 0x20
+#define AHCI_DEV_SIG 0x24
+#define AHCI_DEV_SSTS 0x28
+#define AHCI_DEV_SERR 0x30
+#define AHCI_DEV_SACT 0x34
+#define AHCI_DEV_CI 0x38
+
+#define AHCI_CMD_PRDTL 0x02
+#define AHCI_CMD_PRDBC 0x04
+#define AHCI_CMD_CTBA 0x08
+#define AHCI_CMD_CTBAU 0x0C
+
+#define SATA_SIG_ATA 0x00000101
+#define SATA_SIG_ATAPI 0xEB140101
+#define SATA_SIG_SEMB 0xC33C0101
+#define SATA_SIG_PM 0x96690101
+
+#define HBA_PORT_IPM_ACTIVE 1
+#define HBA_PORT_DET_PRESENT 3
+
+#define ATA_DEV_BUSY 0x80
+#define ATA_DEV_DRQ 0x08
+#define ATA_CMD_READ_DMA_EX 0x25
+#define ATA_CMD_IDENTIFY 0xEC
+
+#define HBA_PxCMD_ST 0x0001
+#define HBA_PxCMD_FRE 0x0010
+#define HBA_PxCMD_FR 0x4000
+#define HBA_PxCMD_CR 0x8000
+#define HBA_PxIS_TFES (1 << 30)
+
+#define LSB(x) ((x)&0xFF)
+#define MSB(x) (((x) >> 8) & 0xFF)
+
+namespace Kernel {
+
+AHCISlot::AHCISlot(VirtualAddress base, AHCIPort& port, u32 index)
+    : m_base_addr(base)
+    , m_index(index)
+{
+    m_dw0 = (DW0*)(m_base_addr.get() + 0x0);
+    m_prdtl = (u16*)(m_base_addr.get() + AHCI_CMD_PRDTL);
+    m_prdbc = (u32*)(m_base_addr.get() + AHCI_CMD_PRDBC);
+    m_ctba = (u32*)(m_base_addr.get() + AHCI_CMD_CTBA);
+    m_ctbau = (u32*)(m_base_addr.get() + AHCI_CMD_CTBAU);
+
+    rebase(port);
+}
+
+void AHCISlot::rebase(AHCIPort& port)
+{
+    *m_prdtl = 8;
+
+    m_ctba_offset = (40 << 10) + (port.index() << 13) + (m_index << 8);
+    *m_ctba = port.base_physical() + m_ctba_offset;
+    *m_ctbau = 0;
+
+    auto table = (u8*)(port.base() + m_ctba_offset);
+    m_cfis = table + 0x00;
+    m_acmd = table + 0x40;
+    m_entries = (PRDT_ENTRY*)(table + 0x80);
+    memset(table, 0, 256);
+}
+
+struct FIS_PIO_SETUP {
+    u8 fis_type; // FIS_TYPE_PIO_SETUP
+
+    u8 pmport : 4; // Port multiplier
```

Comment:
Let's not use bitfields as this is a dangerous way to handle hardware registers. Better to use bitwise AND & OR to handle those bitfields.